ToricVarieties: Chow ring

HereAround opened the pull request.

fingolfin commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,218 @@
+########################
+# Attributes of algebraic cyclees
```
cyclees ?

HereAround replied:
Typo, should be "cycles". Thank you.

fingolfin commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,218 @@
+"""
+function irreducible_components(ac::AlgebraicCycle)::Vector{AlgebraicCycle}
+    if is_irreducible(ac)
+        AlgebraicCycle[ac]
```
I don't understand this line. Do you mean this?
```suggestion
        return [ac]
```

HereAround replied:
Yes, this is what I meant. Thank you!

fingolfin commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,218 @@
+    end
+    expos = matrix(ZZ,[k for k in exponent_vectors(p.f)])
+    indets = gens(parent(p))
+    return AlgebraicCycle[AlgebraicCycle(toric_variety(ac), prod(indets[j]^expos[k,j] for j in 1:ncols(expos))) for k in 1:nrows(expos)]
```
```suggestion
    return [AlgebraicCycle(toric_variety(ac), prod(indets[j]^expos[k,j] for j in 1:ncols(expos))) for k in 1:nrows(expos)]
```
Unless `AlgebraicCycle` is type unstable, it should not be necessary to prescribe the vector type here?

HereAround replied:
I hope it is not type unstable. So I will remove the vector type prescription (and we shall see...).

HereAround commented:
I believe this is ready for review now.

~~Regarding the failure in the doctests, I am not sure if this is related to this PR. (On my local machine, I experience a lot of other doctest-failures, originating from other code in `Oscar`. But this might just be my system... The failure here complaints about a missing literature citation, which - so I believe - was not touched in this PR.)~~ Was my mistake and should be fixed now.

fingolfin commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,218 @@
+"""
+function irreducible_components(ac::AlgebraicCycle)::Vector{AlgebraicCycle}
+    if is_irreducible(ac)
+        [ac]
```
This is essentially a no-op. Did you perhaps mean this?
```suggestion
        return [ac]
```

HereAround replied:
Yes, this was meant. Thank you.

fingolfin replied:
but you did not change it?

HereAround replied:
Strange. I believe I did, but seems the change was lost? I committed this change now.

fingolfin commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
````diff
@@ -0,0 +1,218 @@
+ 7
+```
+"""
+coefficients(ac::AlgebraicCycle) = [coefficient_ring(toric_variety(ac))(k) for k in coefficients(polynomial(ac).f)]
````
So... "the" coefficients of an AlgebraicCycle are not actually well-defined, but depend on some chosen representative `p.f` ?

I think this is not very helpful unless the documentation makes clearer what this is all about? Or is this really the way these objects are treated in the literature?

HereAround replied:
I can see how this could be confusing. I have just added more details.

fingolfin commented on docs/src/ToricVarieties/AlgebraicCycles.md:
```diff
@@ -7,9 +7,68 @@ Pages = ["AlgebraicCycles.md"]
+# The Chow ring and algebraic Cycles
 
-## Special attributes of toric varieties
+Algebraic cycles are elements of the Chow ring.
```
Perhaps you could define what a Chow ring is, or at least sketch it, or give a ref to a source that uses convention that match what you are doing here?

I also noticed that we already have some code dealing with Chow rings, from the Matroids code?. I am totally out of my depth here, but I wonder if these various definition fit together? Perhaps the responsible people could coordinate? CC @bschroter @LukasKuehne 

Some places referencing them include:
- `docs/src/Combinatorics/matroids.md`
- `src/Combinatorics/Matroids/ChowRings.jl`
-

HereAround replied:
Similar to the above. I have added more details. (Aside, this ring is up to isomorphism unique for a toric variety. This is similar in spirit to the `coordinate_ring` of a variety.)

On first glance, I found it surprising to see this for `Matroids` and `Combinatorics.` Then again, I could imagine (not 100% sure though) that one could "strip" the toric variety and define the (or a?) Chow ring just based on combinatorics. This would be similar in spirit to the Stanley-Reisner ideal... But I must say that I have not yet checked in more detail. I would be happy to hear more from the responsible people.

bschroter replied:
@HereAround you are right. One doesn't need a toric variety to define the Chow Ring. The implementation for matroids goes back to the construction of Feichtner and Yuzvinsky and is a key element in all of June Huh's results (for which we received the fields medal).

HereAround replied:
I see. Good to know.

Similar as @fingolfin I wonder then if it would make sense to "marry" the approaches.

To get this started (sorry - I am pretty sure your docu tells but I have not yet checked due to a conference): What information (of the toric variety) do you need to construct the Chow ring?

Right now I take the quotient of the Cox ring by the sum of the Stanley- Reisner and the irrelevant ideal.

bschroter replied:
Probably the best is to first keep it separate and then meet to discuss if/and how one could merge those into one.

HereAround replied:
Ok. I opened an issue (https://github.com/oscar-system/Oscar.jl/issues/1475) so that we do not forget this.

fingolfin commented on docs/src/ToricVarieties/AlgebraicCycles.md:
```diff
@@ -7,9 +7,68 @@ Pages = ["AlgebraicCycles.md"]
+Addition of algebraic cycles `ac1` and `ac2` is implemented by
+`ac1+ac2`. Similarly, we can subtract the cycles by `ac1-ac2`.
+Scalar multiplication with `c` (this could be an integer,
+fmpz or even fmpq number) is supported by `c*ac1`.
```
This is just me and my style, so feel free to ignore me here; but I would move away from fake names for objects ("what is c???") and write something more along these lines (also below). Again, feel free to ignore or do something completely different; at some point we should go over everything and find a common style, but for now we can experiment...

```suggestion
Algebraic cycles can be added and subtracted via the usual `+` and `-`
operators. Moreover, multiplication by scalars from the left is supported
for scalars which are integers or of type `fmpz` or `fmpq`.
```
I've written here "from the left" based on your example snippet, of course adjust this if right scalar multiplication is also supported.

HereAround replied:
That is good. Just need to adjust it for the entire documentation of the toric varieties (likely a lot). Let me look into this...

fingolfin commented on docs/src/ToricVarieties/AlgebraicCycles.md:
```diff
@@ -7,9 +7,68 @@ Pages = ["AlgebraicCycles.md"]
+one can invoke `is_irreducible`.
+
+Equality of algebraic cycles `ac1` and `ac2` is
+implemented by `ac1 == ac2`.
```
Similar to my remark above:
```suggestion
One can check if an algebraic cycle is trivial or irreducible
via `is_trivial` respectively `is_irreducible`.

Equality of algebraic cycles can be tested via `==`.
```
You may also want to turn the reference to `is_trivial` and `is_irreducible` into `@ref` references, though it doesn't really add much here.

However, what *would* be useful if we could somehow convince Documenter to add these places to the index; i.e. find a way so that searching the manual for `is_trivial` also brings up this place. I don't know if Documenter.jl supports that; if it doesn't, we should perhaps file a feature request (or more likely, upvote any existing feature request)... Perhaps @thofma knows?

HereAround replied:
Also a good change of wording. Again, needed for all of the toric varieties.

I suggest to address the `@ref` in a separate PR/issue.

HereAround replied:
Cf. https://github.com/oscar-system/Oscar.jl/issues/1471

fingolfin commented on src/ToricVarieties/AlgebraicCycles/constructors.jl:
```diff
@@ -0,0 +1,196 @@
+            push!(properties_string, "An algebraic cycle on a normal toric variety given by " * string(coeffs[i]) * "*V(" * join(m,",") * ")")
+        else
+            if coeffs[i] > 0
+                push!(properties_string, "+" * string(coeffs[i]) * "*V(" * join(m,",") * ")")
```
This applies in tons of other places: I usually would prefer string interpolation over concatenation. I think it is usually more readable. Well, except for the `join` call here, which I would move outside:
```suggestion
                tmp = join(m,",")
                push!(properties_string, "+$(coeffs[i])V($tmp)")
```

HereAround replied:
Thank you. You indeed mentioned this before, and I just forgot it. Thank you!

fingolfin commented on src/ToricVarieties/AlgebraicCycles/constructors.jl:
```diff
@@ -0,0 +1,196 @@
+        if i == 1
+            push!(properties_string, "An algebraic cycle on a normal toric variety given by " * string(coeffs[i]) * "*V(" * join(m,",") * ")")
+        else
+            if coeffs[i] > 0
```
How about melding these checks:
```suggestion
        elseif coeffs[i] > 0
```
... and then reduce the indentation later and remove the now superfluous `end`?

HereAround replied:
Thank you. That should be great.

fingolfin commented on src/ToricVarieties/AlgebraicCycles/constructors.jl:
```diff
@@ -0,0 +1,196 @@
+
+function Base.:+(ac1::AlgebraicCycle, ac2::AlgebraicCycle)
+    if toric_variety(ac1) !== toric_variety(ac2)
+        throw(ArgumentError("The algebraic cyclees must be defined on identically the same toric variety."))
```
```suggestion
        throw(ArgumentError("The algebraic cycles must be defined on identically the same toric variety"))
```
Typos, and: I think we put o trailing periods into error messages... Not sure if the Julia or Oscar style guides discuss this (@lkastner ?)

HereAround replied:
Thank you.

HereAround committed:
ToricVarieties: Remove trailing periods from error messages

Co-authored-by: Max Horn <max@quendi.de>

HereAround commented:
I have just force pushed. Many of the more "elementary" comments applied to all of the toric varieties and I tried to adjust all instances within the toric context. Points still open are:
- Chow ring(s) appearing in different places within OSCAR",
- The `@ref` might be best addressed separately (cf. https://github.com/oscar-system/Oscar.jl/issues/1471).

benlorenz commented:
Please ignore the failing status for the macos - 1.6 test, the doctests were in fact successful, as shown in the log, sorry about that. (see #1473)

HereAround commented:
Thank you @benlorenz .

HereAround commented:
@lkastner @fingolfin This is ready for a new round of reviews. Thank you!

fingolfin commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,218 @@
+    if iszero(p)
+        return zero(ring)
+    end
+    coeffs = [k for k in coefficients(p.f)]
+    expos = matrix(ZZ,[k for k in exponent_vectors(p.f)])
+    indets = gens(ring)
+    monoms = [prod(indets[j]^expos[k,j] for j in 1:ncols(expos)) for k in 1:nrows(expos)]
+    return sum(coeffs[k]*monoms[k] for k in 1:length(monoms))
```
I think this code is essentially mapping the polynomial `p.f` into the ring `ring`, right? I think we have shorter & more efficient ways for doing this (and if not, we should). I am not at my computer right now and don't want to say something wrong right now (and I am not completely sure what type `p` resp. `p.f` have), but that target ring is of type `MPolyQuo`. Perhaps @fieker can quickly say how one is supposed to convert/map these MPolys

fingolfin replied:
Ah, I think `p` is also an `MPolyQuoElem`... isthe underlying polynomial ring supposed to have the "same" (identical? equivalent? isomorphic?) coefficient ring? Is it a good idea to blindly map variables by their index? I think the documentation for this function should perhaps explain what the word "corresponding" means in the sentence
> Returns the polynomial in `ring` corresponding to the algebraic cycle `ac`.

HereAround replied:
> I think this code is essentially mapping the polynomial `p.f` into the ring `ring`, right? I think we have shorter & more efficient ways for doing this (and if not, we should). I am not at my computer right now and don't want to say something wrong right now (and I am not completely sure what type `p` resp. `p.f` have), but that target ring is of type `MPolyQuo`. Perhaps @fieker can quickly say how one is supposed to convert/map these MPolys

Yes that is the purpose of these lines. I am happy to hear how this can be done better/more efficiently.

HereAround replied:
> Ah, I think `p` is also an `MPolyQuoElem`... isthe underlying polynomial ring supposed to have the "same" (identical? equivalent? isomorphic?) coefficient ring? Is it a good idea to blindly map variables by their index? I think the documentation for this function should perhaps explain what the word "corresponding" means in the sentence
> 
> > Returns the polynomial in `ring` corresponding to the algebraic cycle `ac`.

Ok. I see how this can be confusing. I have added more details here: https://github.com/oscar-system/Oscar.jl/pull/1449/commits/7be42e45269cd42bde286d9d5d7c6fbd41691cab

General philosophy:
- Algebraic cycle: Formal linear combination (over the integers) of irreducible subvarieties. (It makes sense to consider those with rational coefficients too. This is why scalar multiplication with `fmpq` is allowed. But one thing at a time.)
- For a well-defined intersection theory, need to consider them up to a certain equivalence relation. This is the so-called rational equivalence. (Or if you prefer: The ability, to move the cycles in "general position").

For simplicial, complete toric varieties, the algebraic counterpart is roughly speaking as follows:
- Think of algebraic cycles ad polynomials in the Cox ring S (Z^n graded polynomial ring over the complex numbers).
- Think of rational equivalence as taking a quotient of the Cox ring by a certain ideal I (the sum of the Stanley-Reisner ideal and the ideal of linear relations).
- In an abuse of terminology, I use "algebraic cycle" to refer to an equivalence class modulo rational equivalence.
- To get a "feeling" for such an equivalence class, I represent it by an algebraic cycle. In order words, to look at an equivalence class x in S/I I pick y in S such that the projection S -> S/I maps y to x.
- Consequently (as you already pointed out above - thank you for looking it at it carefully!), various functions depend on this choice of y. This includes the functions `coefficients`, `exponents`, `polynomial` as well as the display method.

Is it clear now? If not, I will go over the documentation again, add more details and try to make it better (suggestion welcome).

HereAround replied:
In fact, I have been rushing over a lot of details here. Based on your feedback @fingolfin (thank you for looking at it in detail and asking the right questions), it feels to me that this needs a more fundamental work-over. Let me think about this.

For now, I revert this to a draft.

HereAround committed:
ToricVarieties: Support subvarieties

HereAround committed:
ToricVarieties: Support algebraic cycles

HereAround committed:
PolyhedralGeometry: Implement method for (all) cones of a polyhedral fan

HereAround commented:
@fingolfin @lkastner This is ready for the next round of reviews.

Based on the above feedback, I have:
* Tried to make the distinction between a rational equivalence class and an algebraic cycle crystal clear. 
* As there are relations to closed subvarieties (not yet fully addressed in this PR, but it should be a start), I have extended on this too.

HereAround committed:
ToricVarieties: Implement coefficients of rational equivalence class

HereAround committed:
ToricVarieties: Implement components of rational equivalence class

HereAround committed:
ToricVarieties: Support intersections between

(*) rational equivalence class of algebraic cycles
(*) closed subvariety

considered as elements of the Chow ring

HereAround committed:
ToricVarieties: Execute tests for toric divisor classes

HereAround commented:
Ping @fingolfin  and @lkastner .

lkastner commented on docs/src/ToricVarieties/AlgebraicCycles.md:
````diff
@@ -6,11 +6,142 @@ CurrentModule = Oscar
+```@docs
+representant(ac::RationalEquivalenceClass)
+```
+It can be rather convenient, to investigate such a representant in
````
```suggestion
It can be rather convenient to investigate such a representative in
```

lkastner commented on docs/src/ToricVarieties/AlgebraicCycles.md:
```diff
@@ -6,11 +6,142 @@ CurrentModule = Oscar
+within the regime of simplicial, complete toric varieties this is
+always possible. However, this involves a choice. Consequently,
+the following methods will pick a special choice and return
+values for that particular choice of representant of the rational
```
```suggestion
values for that particular choice of representative of the rational
```

lkastner commented on docs/src/ToricVarieties/AlgebraicCycles.md:
````diff
@@ -6,11 +6,142 @@ CurrentModule = Oscar
+polynomial(ring::MPolyQuo, ac::RationalEquivalenceClass)
+```
+
+### Representant
````
```suggestion
### Representatives
```

lkastner commented on docs/src/ToricVarieties/AlgebraicCycles.md:
````diff
@@ -6,11 +6,142 @@ CurrentModule = Oscar
+equivalence class in question.
+
+```@docs
+representant(ac::RationalEquivalenceClass)
````
```suggestion
representative(ac::RationalEquivalenceClass)
```

lkastner commented on docs/src/ToricVarieties/Subvarieties.md:
```diff
@@ -0,0 +1,49 @@
+a homogeneous ideal in the Cox ring of the toric
+variety in question (cf. proposition 5.2.4 in
+[CLS11](@cite)). As of now, we provide elementary
+support for closd subvarieties of simplicial toric
```
```suggestion
support for closed subvarieties of simplicial toric
```

lkastner commented on docs/src/ToricVarieties/Subvarieties.md:
```diff
@@ -0,0 +1,49 @@
+
+### Special constructors
+
+Empty set and entire variety?
```
```suggestion
```
If there is nothing implemented I would leave that out.

HereAround replied:
Thank you for spotting this. I simply forgot to address this. Let me adjust this (later this afternoon).

HereAround replied:
Just removed this.

lkastner commented on docs/src/ToricVarieties/CohomologyClasses.md:
```diff
@@ -22,31 +22,29 @@ CohomologyClass(l::ToricLineBundle)
+turns out to be (a certain) quotient of the Cox ring. Hence, internally,
+a cohomology class can be thought of as a polynomial in this ring and
+the wedge product corresponds to the product of two (equivalence classes
+of) polynomials.
```
I would suggest to focus more on the implications for the implementation. Maybe rephrase as
```suggestion
The wedge product of cohomology classes is implemented via `*`, 
using internally the multiplication of the corresponding polynomial 
(equivalence classes) in the Cox ring.
```
I am not sure how much of the math we want to explain here, since we have a corresponding book.

HereAround replied:
Fine with me. Thank you for the suggestion!

lkastner commented on docs/src/ToricVarieties/CohomologyClasses.md:
```diff
@@ -22,31 +22,29 @@ CohomologyClass(l::ToricLineBundle)
-Equality of cohomology classes `cc1` and `cc2` is
-implemented by `cc1 == cc2`.
+A cohomology class can be wedged `n`-times with itself via `^n`,
+where `n` can be an integer of of type `fmpz`.
```
`fmpz` sounds like overkill to me, but that is hardly any bottleneck in the computation.

HereAround replied:
I will happy to optimize more. But yes, for now it seems to be no bottle neck.

There is a typo in the text above: "... of of type 'fmpz'". I will fix this later in the afternoon.

HereAround replied:
Typo fixed.

lkastner commented on docs/src/ToricVarieties/ToricDivisorClasses.md:
```diff
@@ -22,37 +22,32 @@ Toric divisor classes are equivalence classes of Weil divisors modulo linear equ
+Equality of toric divisor classes can be tested via `==`.
 
-To check if a toric divisor class `tdc` is trivial, one can invoke `is_trivial(tdc)`.
+To check if a toric divisor class is trivial, one can invoke `is_trivial`.
```
Why not add `is_trivial(ToricDivisor)`? Or is this function undocumented because the `is_trivial` docs grew too large? I remember there was some issue like this.

HereAround replied:
A divisor class is trivial if a divisor representing the class is principal. To check this, we can use `is_principal` for toric divisors.

My expectation for a method `is_trivial` for a toric divisor $D = a_1 D_1 + ... + a_n D_n$ (D_i the torus invariant prime divisors) is to tell if $D \equiv 0$ (i.e. all coefficients $a_i$ vanish). However, this is different from telling if D is a principal divisor.

Do you agree? Does that make sense?

lkastner replied:
It does make sense, but I meant to add it as such a `@docs` block. I think it is fine.

HereAround replied:
Then yes, this is because the `is_trivial` docs grew too large.

Still, we have both functionality and documentation on this (cf. `ToricDivisors.md`):

```
To check if a toric divisor is trivial, one can invoke `is_trivial`.
Internally, this executes the following method:
```@docs
is_principal(td::ToricDivisor)
```

However, I am (on second thought) not really happy with what `is_trivial` does for toric divisors. For divisor classes, this is ok. But for a toric divisor `is_trivial` should act differently. Let me think of a fix for that...

HereAround replied:
@lkastner This is addressed in b035c256d66aa533f29a7d88c4affe65510624dd.

`is_principal` is documented (with @docs block) while `is_trivial` is only mentioned in the documentation.

Note that this seems related to https://github.com/oscar-system/Oscar.jl/issues/1471. 

If I recall correctly, I could not simply add an @ref for `is_trivial` because somewhere in `OSCAR` there was (at the time that I tried = several weeks ago) documentation for an `is_trivial` method, which led to some error... Maybe this is best addressed separately?

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+
+
+@doc Markdown.doc"""
+    representant(ac::RationalEquivalenceClass)
```
```suggestion
    representative(ac::RationalEquivalenceClass)
```

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+julia> ac*ac
+A rational equivalence class on a normal toric variety represented by 34V(x2,x3)
+
+julia> representant(ac*ac)
```
```suggestion
julia> representative(ac*ac)
```

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
````diff
@@ -0,0 +1,263 @@
+34*x2*x3
+```
+"""
+@attr MPolyElem_dec{fmpq, fmpq_mpoly} function representant(ac::RationalEquivalenceClass)
````
```suggestion
@attr MPolyElem_dec{fmpq, fmpq_mpoly} function representative(ac::RationalEquivalenceClass)
```

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+    mapped_monomials = [map_gens_of_chow_ring_to_cox_ring(toric_variety(ac))[m] for m in monomials(polynomial(ac).f)]
+    return sum([coeffs[i]*mapped_monomials[i] for i in 1:length(mapped_monomials)])
+end
+export representant
```
```suggestion
export representative
```

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+
+A rational equivalence class of algebraic cycles can
+be represented by a polynomial in the Cox ring (cf.
+function `representant`). By no means is this
```
```suggestion
function `representative`). By no means is this
```

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+A rational equivalence class of algebraic cycles can
+be represented by a polynomial in the Cox ring (cf.
+function `representant`). By no means is this
+representant unique. For one such choice of polynomial
```
```suggestion
representative unique. For one such choice of polynomial
```

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+formal linear sum, with the coefficients computed by#
+the method `coefficients(ac::RationalEquivalenceClass)`,
+defines an algebraic cycle, whose rational equivalence
+class is identical to the one given to this method.
```
```suggestion
Turn each monomial
of `representative(ac)` into a closed subvariety and return
the list formed from these subvarieties. Note that
each of these subvarieties is irreducible and their
formal linear sum, with the coefficients computed by#
the method `coefficients(ac::RationalEquivalenceClass)`,
defines an algebraic cycle, whose rational equivalence
class is identical to the one given to this method.
```
I'd like to avoid duplication even in the docs.

HereAround replied:
Yes, that is better. Thank you!

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+    end
+    variety = toric_variety(ac)
+    gs = gens(cox_ring(toric_variety(ac)))
+    mons = [m for m in monomials(representant(ac))]
```
```suggestion
    mons = [m for m in monomials(representative(ac))]
```

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+function `representant`). By no means is this
+representant unique. For one such choice of polynomial
+in the Cox ring, this method returns the coefficients
+of said polynomial.
```
Why not just
```suggestion
Return the coefficients of `polynomial(ac)`.
```
?

HereAround replied:
That is perfect. Thank you!

HereAround replied:
Cannot apply this suggestion (presumably another change above modified this code segment). Let me address this later this afternoon.

HereAround replied:
Change applied.

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+export toric_variety
+
+
+@doc Markdown.doc"""
+    polynomial(ac::RationalEquivalenceClass)
+
+On a simplicial and complete toric variety, the Chow ring
+is isomorphic to a certain quotient of the Cox ring. This
+function returns the ring element corresponding to a given
+rational equivalence class of algebraic cycles.
```
What is the difference between `representative` and `polynomial`? Is it only the ring?

HereAround replied:
The rational for representative is to present a rational equivalence class as a formal linear sum of complete intersections. The latter must of course not involve self-intersection loci. This is the key difference between the two. Explicitly:

1. `polynomial` belongs to the Chow ring, `representative` to the Cox ring.
2. `polynomial` can involve monomials, which you would interpret as self-intersections. For example $2 \cdot x^2$ corresponds to 2-times the self-intersection locus of the toric divisor $V(x)$ with itself. For `representative` I will pick a rationally equivalent algebraic cycle which only involves complete intersections. For the polynomial in question this means that in every monomial no variable appears to a power higher than $1$. For example, instead of $2 \cdot x^2$ it could be $-2 \cdot y \cdot z$, which you can read as $-2$ times the complete intersection $V(y,z)$.

(The reason for 'representative' sitting in the Cox ring is that any closed subvariety of a "nice" (I do not remember the conditions right now) toric variety is always cut-out as the vanishing locus of a finite number of polynomials in the Cox ring. So it seemed natural to put the representative in the Cox ring. But this is not strictly necessary.)

lkastner commented on src/ToricVarieties/AlgebraicCycles/attributes.jl:
```diff
@@ -0,0 +1,263 @@
+Returns a polynomial in the Cox ring which represents
+a rational equivalence class of algebraic cycles without
+self-intersections. Note that this is by no means unique,
+i.e. the returned polynomial involves a choice.
```
Why not just
```suggestion
Return a polynomial in the Cox ring mapping to `polynomial(ac)`.
```
?

HereAround replied:
Fine with me.

lkastner commented on src/ToricVarieties/AlgebraicCycles/constructors.jl:
```diff
@@ -0,0 +1,304 @@
+
+function Base.:+(ac1::RationalEquivalenceClass, ac2::RationalEquivalenceClass)
+    if toric_variety(ac1) !== toric_variety(ac2)
+        throw(ArgumentError("The rational equivalence classes must be defined on identically the same toric variety"))
```
```suggestion
        throw(ArgumentError("The rational equivalence classes must be defined on the same toric variety"))
```

lkastner commented on src/ToricVarieties/AlgebraicCycles/constructors.jl:
```diff
@@ -0,0 +1,304 @@
+        join(io, "A trivial rational equivalence class on a normal toric variety")
+    else
+      # otherwise, extract properties to represent the rational equivalence class
+      r = representant(ac)
```
```suggestion
      r = representative(ac)
```

lkastner commented on test/ToricVarieties/runtests.jl:
```diff
@@ -552,3 +554,78 @@ end
+@testset "Attributes of rational equivalence cycles" begin
+    @test dim(toric_variety(ac1)) == 2
+    @test polynomial(ac1) == 0
+    @test parent(representant(ac1)) == cox_ring(toric_variety(ac1))
```
```suggestion
    @test parent(representative(ac1)) == cox_ring(toric_variety(ac1))
```

lkastner commented on src/ToricVarieties/cohomCalg/auxilliary.jl:
```diff
@@ -62,7 +62,7 @@ function contributing_denominators(variety::AbstractNormalToricVariety)
     # ambiguous monomial contributions found during execution?
     if last(output_string_reduced) != "There are no ambiguous contribution monomials to the variety."
-        throw(ArgumentError("cohomCalg experienced ambiguous monomial contributions."))
+        throw(ArgumentError("cohomCalg experienced ambiguous monomial contributions"))
```
Maybe something like
```suggestion
        throw(ArgumentError("cohomCalg encountered ambiguous monomial contributions"))
```
? "experienced" to me sounds like cohomCalg developing feelings. ;)

HereAround replied:
If my recollection is correct, then I copied the error message that `cohomCalg` would print out. In any case, I agree that "encountered" is better than "experienced".

lkastner commented on src/ToricVarieties/CohomologyClasses/special_attributes.jl:
```diff
@@ -17,7 +17,7 @@ julia> ngens(cohomology_ring(p2))
 @attr MPolyQuo function cohomology_ring(v::AbstractNormalToricVariety)
     if !is_simplicial(v) || !is_complete(v)
-        throw(ArgumentError("The cohomology ring is (currently) only supported for simplicial and complete toric varieties."))
+        throw(ArgumentError("The cohomology ring is (currently) only supported for simplicial and complete toric varieties"))
```
```suggestion
        throw(ArgumentError("The cohomology ring is only supported for simplicial and complete toric varieties"))
```

lkastner reviewed (approved):
- "identically the same": Maybe you could clarify differently? Like "must be defined on the same toric variety, i.e. the same OSCAR variable" or something similar.
- Some explanations of functions are rather exhaustive, while others are quite short. It is not quite clear to me what distinguishes one kind of functions from the other.
- Since you added a lot of error messages you could think about testing them. Look for `@test_throws` in the other tests for examples.
- At some point there should be some decision on how to handle one-line functions (not by us). Here they are all exhaustively documented which requires many times the effort to write and maintain than the implementation itself. I think this will be infeasible in the long run.

HereAround commented:
Thank you @lkastner for the many suggestions. Most of which were applied. Some require extra care, which I hope to address towards late afternoon (Philly). Also, I believe it makes sense to squash your many suggestions?

> * "identically the same": Maybe you could clarify differently? Like "must be defined on the same toric variety, i.e. the same OSCAR variable" or something similar.
> 

Yes that is meant. I just wanted to emphasize that an abstract isomorphism is not enough (at least for now) but it must be the same OSCAR variable. Your suggestion makes that clearer. So probably worth adjusting everywhere (= many many instances) within the toric varieties.


>     * Some explanations of functions are rather exhaustive, while others are quite short. It is not quite clear to me what distinguishes one kind of functions from the other.
> 

@fingolfin asked for some details above, so I tried to explain those more carefully. (Rational: Whatever @fingolfin picks up upon, others might too. So the hope is, that the explanations are longer for complicated material and shorter for simple content. This is of course only a hope.)

>     * Since you added a lot of error messages you could think about testing them. Look for `@test_throws` in the other tests for examples.
> 

I agree. This is why I added the error tests for the toric varieties. But it might not cover all warnings yet. Let me check in more detail later which ones are not yet covered. Then I can extend.

>     * At some point there should be some decision on how to handle one-line functions (not by us). Here they are all exhaustively documented which requires many times the effort to write and maintain than the implementation itself. I think this will be infeasible in the long run.

Ok.

HereAround committed:
ToricVarieties: Code improvement

Co-authored-by: Lars Kastner <lkastner@users.noreply.github.com>

HereAround commented on test/ToricVarieties/runtests.jl:
```diff
@@ -552,3 +554,78 @@ end
+ac5 = RationalEquivalenceClass(CohomologyClass(dP3, x3))
+ac6 = RationalEquivalenceClass(ToricLineBundle(ntv,[1,1]))
+
+@testset "Error of rational equivalence classes" begin
+  @test_throws ArgumentError RationalEquivalenceClass(antv, [1,2,3])
+  @test_throws ArgumentError RationalEquivalenceClass(toric_variety(ac1),[1,2,3])
+  @test_throws ArgumentError ac1 + ac3
+  @test_throws ArgumentError ac1 - ac3
+  @test_throws ArgumentError ac1 * ac3
+  @test_throws ArgumentError sv1 * ac1
+  @test_throws ArgumentError ac2 * sv1
+  @test_throws ArgumentError sv1 * sv3
+end
```
@lkastner I believe these tests cover most (if not all?) of the error messages that I added in this PR. Do you agree? If not, which errors need testing?

lkastner replied:
Sorry, I missed these. These are perfect.

HereAround commented:
@lkastner Up to the 3 comments above that I left open, I have integrated all of your suggestions.

lkastner commented:
> Also, I believe it makes sense to squash your many suggestions?

Seems like you already did that?

HereAround commented:
> > Also, I believe it makes sense to squash your many suggestions?
> 
> Seems like you already did that?

Indeed, I have squashed them. (Were order 30 individual small commits otherwise. I hope you don't mind?)

HereAround committed:
ToricVarieties: Properly implement is_trivial for toric divisors
